src/coreComponents/constitutive/fluid/BlackOilFluid.cpp

```diff
@@ -31,76 +26,443 @@ using namespace dataRepository;
+    localIndex j = i + 1;
+    while( oilTable[j].size() == 3 )
+    {
+      m_PVTO.m_undersaturatedPressure[iSat].emplace_back( oilTable[j][0] - m_PVTO.m_bubblePressure[iSat] );
+      m_PVTO.m_undersaturatedBo[iSat].emplace_back( oilTable[j][1] );
+      m_PVTO.m_undersaturatedViscosity[iSat].emplace_back( oilTable[j][2] );
+      branchSize++;
+      j++;
+      if( j == oilTable.size())
+      {
+        break;
+      }
+    }
+    i = i + branchSize;
+    iSat++;
+  }
+
+  // add 1 atm value if does not exist yet
+  if( !isZero( m_PVTO.m_Rs[0] ) )
+  {
+    real64 const refPressure = 101325.0;
+    real64 const viscosity = logExtrapolation( m_PVTO.m_bubblePressure[1],
+                                               m_PVTO.m_saturatedViscosity[1],
+                                               m_PVTO.m_bubblePressure[0],
+                                               m_PVTO.m_saturatedViscosity[0],
+                                               refPressure );
+
+    m_PVTO.m_Rs.emplace( 0, 0. );
+    m_PVTO.m_bubblePressure.emplace( 0, refPressure );
+    m_PVTO.m_saturatedBo.emplace( 0, 1.0 );
+    m_PVTO.m_saturatedViscosity.emplace( 0, viscosity );
+    m_PVTO.m_nSaturatedPoints++;
 
-template< typename ARRAY >
-void checkInputSize( ARRAY const & array, localIndex const expected, string const & attr )
+    array1d< real64 > tmp( 1 );
+    tmp[0] = 0.;
+    m_PVTO.m_undersaturatedPressure.emplace( 0, tmp );
+    tmp[0] = 1.;
+    m_PVTO.m_undersaturatedBo.emplace( 0, tmp );
+    tmp[0] = viscosity;
+    m_PVTO.m_undersaturatedViscosity.emplace( 0, tmp );
+  }
+
+
+  /// find the max relative pressure (later used to extend the tables)
+  m_PVTO.m_maxRelativePressure = 0;
+
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
+    real64 const maxP = *( std::max_element( m_PVTO.m_undersaturatedPressure[i].begin(), m_PVTO.m_undersaturatedPressure[i].end() ) );
+    m_PVTO.m_maxRelativePressure = std::max( maxP, m_PVTO.m_maxRelativePressure );
+  }
+
+  // densities
+  m_PVTO.m_surfaceMassDensity.resize( NC_BO-1 );
+  m_PVTO.m_surfaceMoleDensity.resize( NC_BO-1 );
+
+  using PT = BlackOilFluid::PhaseType;
+
+  m_PVTO.m_surfaceMassDensity[PT::OIL] = oilSurfaceMassDensity;
+  m_PVTO.m_surfaceMoleDensity[PT::OIL] = oilSurfaceMassDensity / oilSurfaceMolecularWeight;
+
+  m_PVTO.m_surfaceMassDensity[PT::GAS] = gasSurfaceMassDensity;
+  m_PVTO.m_surfaceMoleDensity[PT::GAS] = gasSurfaceMassDensity / gasSurfaceMolecularWeight;
+
+}
+
+void BlackOilFluid::createUndersaturatedProperties()
 {
-  GEOSX_THROW_IF_NE_MSG( array.size(), expected,
-                         "BlackOilFluid: invalid number of entries in " << attr << " attribute",
-                         InputError );
+  localIndex upperBranchIndex = m_PVTO.m_nSaturatedPoints - 1;
+
+  // construct undersaturated branches by interpolation, start from second highest Rs
+
+  for( localIndex iCurrent = m_PVTO.m_nSaturatedPoints - 2; iCurrent-- > 0; )
```

Comment:
Ah, the famous ["goes to" operator](https://stackoverflow.com/questions/1642028/what-is-the-operator-in-c-c)! I find this hard to process personally, something like
```suggestion
  for( localIndex iCurrent = m_PVTO.m_nSaturatedPoints - 2; iCurrent >= 0; --iCurrent )
```
is more natural :)